File: code_samples/section13/example_3_auto_complete/trie_autocomplete.java
package code_samples.section13.example_3_auto_complete;

import java.io.BufferedReader;   // Efficient line-by-line reading of text files
import java.io.FileReader;       // Character stream for reading a file
import java.io.IOException;      // Exception type for I/O failures
import java.util.ArrayList;      // Dynamic list for returning autocomplete results

/*
 * Trie Autocomplete (Java)
 *
 * This example demonstrates a trie (prefix tree) that supports:
 * - Inserting words (case-insensitive)
 * - Autocompleting a prefix by returning up to N suggestions
 *
 * Key behavior / constraints:
 * - Only letters 'a'–'z' are supported (case-insensitive).
 * - If any character in a word/prefix is not a–z:
 *   - insert() rejects the entire word
 *   - walk() returns null
 *   - autocomplete() returns an empty list
 *
 * Autocomplete ordering:
 * - Suggestions are generated in lexicographic order because DFS visits
 *   children in 'a'..'z' order.
 *
 * Dictionary loading:
 * - Default dictionary path is "..\\data\\words.txt"
 * - Relative paths resolve from the working directory where Java runs.
 */
public class trie_autocomplete {

    /*
     * Number of supported lowercase English letters.
     */
    private static final int ALPHABET_SIZE = 26;

    /*
     * Maximum word length constant (present for parity with other examples).
     * Not used in this implementation because StringBuilder grows dynamically.
     */
    @SuppressWarnings("unused")
    private static final int WORD_MAX = 128;

    /*
     * Trie node definition.
     *
     * Each node contains:
     * - children: references to child nodes for each letter 'a'–'z'
     * - isEndOfWord: indicates whether a complete inserted word ends at this node
     */
    private static class TrieNode {
        TrieNode[] children = new TrieNode[ALPHABET_SIZE]; // Child pointers
        boolean isEndOfWord = false;                       // End-of-word marker
    }

    /*
     * Root node of the trie.
     * Always exists and does not correspond to any particular character.
     */
    private final TrieNode root = new TrieNode();

    /*
     * Convert a lowercase character into an array index.
     *
     * Parameters:
     * - c: expected to be between 'a' and 'z'
     *
     * Returns:
     * - 0..25 if valid
     * - -1 if invalid
     */
    private static int index(char c) {
        if (c < 'a' || c > 'z') return -1;
        return c - 'a';
    }

    /*
     * Insert a word into the trie.
     *
     * Behavior:
     * - Normalizes to lowercase
     * - Rejects the entire word if any character is not 'a'–'z'
     * - Creates nodes lazily as needed
     * - Marks the final node as an end-of-word
     *
     * Parameters:
     * - word: word to insert
     */
    public void insert(String word) {
        TrieNode current = root;

        // Traverse each character of the word
        for (int i = 0; i < word.length(); i++) {
            char c = Character.toLowerCase(word.charAt(i));
            int idx = index(c);

            // Reject word if any character is invalid
            if (idx < 0) return;

            // Allocate child node if missing
            if (current.children[idx] == null) current.children[idx] = new TrieNode();

            // Advance in the trie
            current = current.children[idx];
        }

        // Mark the final node as representing a complete word
        current.isEndOfWord = true;
    }

    /*
     * Walk the trie according to a prefix and return the final node.
     *
     * Behavior:
     * - Lowercases each character
     * - Returns null if prefix contains invalid characters or missing path
     *
     * Parameters:
     * - prefix: prefix to traverse
     *
     * Returns:
     * - TrieNode at the end of the prefix, or null if not found/invalid
     */
    private TrieNode walk(String prefix) {
        TrieNode current = root;

        // Traverse each character of the prefix
        for (int i = 0; i < prefix.length(); i++) {
            char c = Character.toLowerCase(prefix.charAt(i));
            int idx = index(c);

            // Invalid character -> prefix not supported
            if (idx < 0) return null;

            // Missing path -> prefix not present in trie
            if (current.children[idx] == null) return null;

            // Advance in the trie
            current = current.children[idx];
        }

        return current;
    }

    /*
     * Return up to `limit` autocomplete suggestions for the given prefix.
     *
     * Steps:
     * 1) Find the prefix node using walk(prefix).
     * 2) Seed a StringBuilder with the normalized (lowercased) prefix.
     * 3) DFS from the prefix node, appending letters to the buffer as we explore.
     * 4) Add completed words (end-of-word nodes) to the output list.
     *
     * Parameters:
     * - prefix: prefix to autocomplete
     * - limit: maximum number of suggestions to return
     *
     * Returns:
     * - ArrayList<String> of suggestions, possibly empty
     *
     * Notes:
     * - The prefix itself is included in results only if it is also a complete word.
     */
    public ArrayList<String> autocomplete(String prefix, int limit) {
        TrieNode start = walk(prefix);

        // Output list (returned even when empty)
        ArrayList<String> out = new ArrayList<>();

        // If prefix is invalid or not found, return empty list
        if (start == null) return out;

        // Seed buffer with the lowercased prefix; extra capacity helps reduce reallocations
        StringBuilder buf = new StringBuilder(prefix.length() + 16);
        for (int i = 0; i < prefix.length(); i++) {
            buf.append(Character.toLowerCase(prefix.charAt(i)));
        }

        // DFS beneath the prefix node to collect suggestions
        dfsCollect(start, buf, out, limit);
        return out;
    }

    /*
     * Depth-first traversal to collect autocomplete suggestions.
     *
     * Parameters:
     * - node: current subtree root
     * - buf: mutable buffer holding the current word (prefix + explored suffix)
     * - out: output list of suggestions
     * - limit: maximum number of suggestions to collect
     *
     * Behavior:
     * - If node.isEndOfWord is true, add buf.toString() as a suggestion.
     * - Visit children in 'a'..'z' order to produce lexicographic results.
     * - Backtrack by removing the last character after each recursive call.
     * - Stop early once out.size() reaches limit.
     */
    private void dfsCollect(TrieNode node, StringBuilder buf, ArrayList<String> out, int limit) {
        // Base cases: nothing to explore, or we've already collected enough results
        if (node == null || out.size() >= limit) return;

        // If this node terminates a word, capture the current buffer as a suggestion
        if (node.isEndOfWord) {
            out.add(buf.toString());
            if (out.size() >= limit) return;
        }

        // Explore children in lexicographic order
        for (int i = 0; i < ALPHABET_SIZE; i++) {
            if (node.children[i] != null) {
                // Append this character to the current word buffer
                buf.append((char)('a' + i));

                // Recurse into the child subtree
                dfsCollect(node.children[i], buf, out, limit);

                // Backtrack: remove the last appended character
                buf.deleteCharAt(buf.length() - 1);

                // Stop early if limit reached
                if (out.size() >= limit) return;
            }
        }
    }

    /*
     * Load words from a dictionary file into the trie.
     *
     * Parameters:
     * - trie: trie instance to populate
     * - path: dictionary file path
     *
     * Returns:
     * - number of non-empty trimmed lines processed (insert attempted for each)
     * - 0 if the file cannot be opened/read
     *
     * Notes:
     * - insert() rejects words with invalid characters, but this loader still
     *   increments count for each non-empty trimmed line (as written).
     */
    private static int loadDictionary(trie_autocomplete trie, String path) {
        int count = 0;

        // Try-with-resources ensures the reader closes even if an exception occurs
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            String line;

            // Read file line-by-line
            while ((line = br.readLine()) != null) {
                line = line.trim();

                // Skip empty lines
                if (line.isEmpty()) continue;

                // Insert word into trie (may be rejected if invalid)
                trie.insert(line);
                count++;
            }
        } catch (IOException e) {
            // Report file access / I/O errors
            System.err.println("Failed to open dictionary: " + path);
            System.err.println(e.getMessage());
            return 0;
        }

        return count;
    }

    /*
     * Parse a positive integer from a string, or return a default.
     *
     * Parameters:
     * - s: input string representing an integer
     * - def: default to return if parsing fails or value is non-positive
     *
     * Returns:
     * - parsed value if > 0
     * - def otherwise
     */
    private static int parseIntOrDefault(String s, int def) {
        try {
            int v = Integer.parseInt(s);
            return v > 0 ? v : def;
        } catch (Exception e) {
            return def;
        }
    }

    /*
     * Program entry point.
     *
     * Usage:
     * - args[0] (optional): dictionary path (default: "..\\data\\words.txt")
     * - args[1] (optional): prefix (default: "ab")
     * - args[2] (optional): limit (default: 20)
     *
     * Demonstrates:
     * - Dictionary loading
     * - Autocomplete query for a prefix
     * - Printing returned suggestions
     */
    public static void main(String[] args) {
        trie_autocomplete trie = new trie_autocomplete();

        // Default path matches lesson convention; allow override via CLI args
        String dictPath = (args.length > 0) ? args[0] : "..\\data\\words.txt";
        String prefix   = (args.length > 1) ? args[1] : "ab";
        int limit       = (args.length > 2) ? parseIntOrDefault(args[2], 20) : 20;

        // Load dictionary into trie
        int loaded = loadDictionary(trie, dictPath);
        System.out.println("Loaded " + loaded + " words from " + dictPath);
        System.out.println();

        // Run autocomplete and print results
        System.out.println("Autocomplete(\"" + prefix + "\") [limit=" + limit + "]");
        for (String w : trie.autocomplete(prefix, limit)) {
            System.out.println(w);
        }
    }
}

File: code_samples/section13/example_3_auto_complete/trie_autocomplete.cs
// Trie Autocomplete (top-level, dotnet-script friendly)
//
// This script implements a trie (prefix tree) that supports autocomplete.
// It is designed to run as a single-file, top-level C# script using dotnet-script.
//
// Key behavior / constraints:
// - Only letters 'a'–'z' are accepted (case-insensitive).
// - If any character in a word is not a–z, Insert() rejects the entire word.
// - If any character in a prefix is not a–z, Walk() returns null and
//   Autocomplete() returns an empty list.
// - Autocomplete results are produced in lexicographic order because DFS
//   visits children in 'a'..'z' order.
//
// Dictionary loading:
// - Loads from a file path (default: ..\data\words.txt)
// - Relative paths resolve from the working directory where the script runs.

const int ALPHABET_SIZE = 26; // Number of supported lowercase English letters

// Root node of the trie; the entry point for all insert/search/autocomplete operations
TrieNode root = new();

/*
 * Convert a lowercase character into an index [0, 25].
 *
 * Parameters:
 * - c: expected to be 'a'..'z'
 *
 * Returns:
 * - 0..25 if valid
 * - -1 if invalid
 */
int Index(char c) => (c >= 'a' && c <= 'z') ? (c - 'a') : -1;

/*
 * Insert a word into the trie.
 *
 * Behavior:
 * - Normalizes characters to lowercase (case-insensitive)
 * - Rejects the entire word if any character is not 'a'–'z'
 * - Creates missing nodes as needed
 * - Marks the final node as an end-of-word
 *
 * Parameters:
 * - word: string to insert
 */
void Insert(string word)
{
    TrieNode current = root;

    // Traverse each character in the word
    foreach (char raw in word)
    {
        // Normalize character to lowercase (invariant) for consistent indexing
        char c = char.ToLowerInvariant(raw);

        // Convert to index
        int idx = Index(c);

        // Reject whole word if any invalid character appears
        if (idx < 0) return;

        // Allocate child node if missing
        if (current.Children[idx] == null) current.Children[idx] = new TrieNode();

        // Move to the next trie node
        current = current.Children[idx];
    }

    // Mark the final node as a complete word
    current.IsEndOfWord = true;
}

/*
 * Walk the trie for a given prefix and return the node at the end of the prefix.
 *
 * Behavior:
 * - Lowercases each character
 * - Returns null if any character is invalid (non a–z)
 * - Returns null if any required child node is missing (prefix not found)
 *
 * Parameters:
 * - prefix: prefix to walk
 *
 * Returns:
 * - TrieNode representing the final character of prefix, or null if not found/invalid
 */
TrieNode Walk(string prefix)
{
    TrieNode current = root;

    foreach (char raw in prefix)
    {
        char c = char.ToLowerInvariant(raw);
        int idx = Index(c);

        // Invalid character -> prefix not supported
        if (idx < 0) return null;

        // Missing path -> prefix not present in trie
        if (current.Children[idx] == null) return null;

        current = current.Children[idx];
    }

    return current;
}

/*
 * Depth-first traversal to collect autocomplete suggestions.
 *
 * Parameters:
 * - node: subtree root to explore
 * - outWords: output list of collected words
 * - buffer: character buffer containing the current word under construction
 * - depth: current length (number of valid chars in buffer)
 * - limit: maximum number of suggestions to collect
 *
 * Behavior:
 * - When node.IsEndOfWord is true, the current buffer[0..depth) is a complete word.
 * - Visits children in 'a'..'z' order to produce lexicographic suggestions.
 * - Stops early once outWords.Count reaches limit.
 *
 * Implementation detail:
 * - This DFS writes the next letter into buffer[depth] before recursing.
 * - The buffer is reused across recursion to avoid repeated allocations.
 */
void DfsCollect(TrieNode node, List<string> outWords, char[] buffer, int depth, int limit)
{
    // Base cases: no subtree, or we've reached the desired limit
    if (node == null || outWords.Count >= limit) return;

    // If this node represents the end of a word, add the current buffer as a suggestion
    if (node.IsEndOfWord)
    {
        outWords.Add(new string(buffer, 0, depth));
        if (outWords.Count >= limit) return;
    }

    // Explore children in lexicographic order
    for (int i = 0; i < ALPHABET_SIZE; i++)
    {
        if (node.Children[i] != null)
        {
            // Append the next character to the buffer
            buffer[depth] = (char)('a' + i);

            // Recurse deeper to extend the word
            DfsCollect(node.Children[i], outWords, buffer, depth + 1, limit);

            // Stop immediately once we reach the limit (avoid extra traversal)
            if (outWords.Count >= limit) return;
        }
    }
}

/*
 * Return autocomplete suggestions for a prefix, up to a given limit.
 *
 * Steps:
 * 1) Find the trie node corresponding to the prefix via Walk().
 * 2) Seed a character buffer with the (lowercased) prefix.
 * 3) DFS from the prefix node to enumerate suffixes and collect full words.
 *
 * Parameters:
 * - prefix: prefix to autocomplete
 * - limit: maximum number of suggestions to return
 *
 * Returns:
 * - List of suggestions (possibly empty)
 *
 * Notes:
 * - The returned suggestions include the prefix itself only if the prefix
 *   is also a complete word (i.e., start.IsEndOfWord is true).
 */
List<string> Autocomplete(string prefix, int limit)
{
    TrieNode start = Walk(prefix);

    // If prefix is invalid or not found, return empty result
    if (start == null) return [];

    // Pre-size list capacity to avoid repeated growth allocations
    var outWords = new List<string>(limit);

    // Buffer used to build candidate words (prefix + explored suffix letters)
    // Size 256 is a fixed cap for suggestions; longer words will be truncated by seeding logic.
    char[] buffer = new char[256];

    // Seed the buffer with the prefix (lowercased) and track initial depth
    int depth = 0;
    foreach (char raw in prefix)
    {
        if (depth >= buffer.Length) break;         // Prevent buffer overflow
        buffer[depth++] = char.ToLowerInvariant(raw);
    }

    // Collect suggestions under the prefix node
    DfsCollect(start, outWords, buffer, depth, limit);
    return outWords;
}

/*
 * Load a dictionary file into the trie.
 *
 * Parameters:
 * - path: filesystem path to the dictionary file
 *
 * Returns:
 * - number of non-empty trimmed lines processed (Insert attempted for each)
 * - 0 if the file cannot be opened/read
 *
 * Notes:
 * - Insert() may reject words containing invalid characters, but this function
 *   still increments "count" for each non-empty trimmed line (as written).
 */
int LoadDictionary(string path)
{
    int count = 0;

    try
    {
        // Streams lines lazily (better for large files than ReadAllLines)
        foreach (var line in File.ReadLines(path))
        {
            var w = line.Trim();
            if (w.Length == 0) continue;

            Insert(w);
            count++;
        }
    }
    catch (Exception ex)
    {
        // Report file access / I/O errors
        Console.Error.WriteLine($"Failed to open dictionary: {path}");
        Console.Error.WriteLine(ex.Message);
        return 0;
    }

    return count;
}

/*
 * Parse an integer from a string or return a default.
 *
 * Parameters:
 * - s: input string (possibly null/empty/whitespace)
 * - def: fallback default value
 *
 * Returns:
 * - parsed positive integer if valid (> 0)
 * - def otherwise
 */
int ParseIntOrDefault(string s, int def)
{
    if (string.IsNullOrWhiteSpace(s)) return def;
    return int.TryParse(s, out int v) && v > 0 ? v : def;
}

// -------------------- Main --------------------
// Usage: dotnet script .\trie_autocomplete.cs [dictPath] [prefix] [limit]

/*
 * Resolve dictionary path and optional arguments:
 * - Args[0]: dictionary path
 * - Args[1]: prefix
 * - Args[2]: limit
 */
string dictPath = (Args.Count > 0) ? Args[0] : @"..\data\words.txt";
string prefix   = (Args.Count > 1) ? Args[1] : "ab";
int limit       = (Args.Count > 2) ? ParseIntOrDefault(Args[2], 20) : 20;

// Load dictionary words into the trie
int loaded = LoadDictionary(dictPath);
Console.WriteLine($"Loaded {loaded} words from {dictPath}\n");

// Run autocomplete and print the results
Console.WriteLine($"Autocomplete(\"{prefix}\") [limit={limit}]");
foreach (var w in Autocomplete(prefix, limit))
    Console.WriteLine(w);

Console.WriteLine();

// Helpful diagnostics for debugging relative path resolution
Console.WriteLine($"Working directory = {Directory.GetCurrentDirectory()}");
Console.WriteLine($"Resolved dictionary path = {Path.GetFullPath(dictPath)}");

/*
 * Trie node definition.
 *
 * Each node contains:
 * - Children: 26 child references for 'a'–'z'
 * - IsEndOfWord: true if a complete word ends at this node
 *
 * This variant does not store prefix counts because autocomplete enumerates
 * words via DFS instead.
 */
class TrieNode
{
    public TrieNode[] Children = new TrieNode[ALPHABET_SIZE]; // Child nodes
    public bool IsEndOfWord = false;                          // End-of-word marker
}
